src/sagemaker_xgboost_container/algorithm_mode/serve_utils.py

```diff
@@ -0,0 +1,254 @@
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import csv
+import io
+import json
+import numpy as np
+import os
+
+from sagemaker_containers.record_pb2 import Record
+from sagemaker_containers._recordio import _write_recordio
+
+from sagemaker_xgboost_container.algorithm_mode import integration
+
+
+logging = integration.setup_main_logger(__name__)
+
+SAGEMAKER_INFERENCE_OUTPUT = 'SAGEMAKER_INFERENCE_OUTPUT'
+
+# xgboost objective learning tasks
+# https://xgboost.readthedocs.io/en/release_0.90/parameter.html#learning-task-parameters
+REG_SQUAREDERR = 'reg:squarederror'
+REG_LOG = 'reg:logistic'
+REG_GAMMA = 'reg:gamma'
+REG_TWEEDIE = 'reg:tweedie'
+BINARY_LOG = 'binary:logistic'
+BINARY_LOGRAW = 'binary:logitraw'
+BINARY_HINGE = 'binary:hinge'
+MULTI_SOFTMAX = 'multi:softmax'
+MULTI_SOFTPROB = 'multi:softprob'
+
+
+# classification selectable inference keys
+PREDICTED_LABEL = 'predicted_label'
+LABELS = 'labels'
+PROBABILITY = 'probability'
+PROBABILITIES = 'probabilities'
+RAW_SCORE = 'raw_score'
+RAW_SCORES = 'raw_scores'
+
+# regression selectable inference keys
+PREDICTED_SCORE = 'predicted_score'
+
+# all supported selecable content keys
+ALL_VALID_SELECT_KEYS = [PREDICTED_LABEL, LABELS, PROBABILITY, PROBABILITIES, RAW_SCORE, RAW_SCORES, PREDICTED_SCORE]
+
+# mapping of xgboost objective functions to valid selectable inference content
+VALID_OBJECTIVES = {
+    REG_SQUAREDERR: [PREDICTED_SCORE],
+    REG_LOG: [PREDICTED_SCORE],
+    REG_GAMMA: [PREDICTED_SCORE],
+    REG_TWEEDIE: [PREDICTED_SCORE],
+    BINARY_LOG: [PREDICTED_LABEL, LABELS, PROBABILITY, PROBABILITIES, RAW_SCORE, RAW_SCORES],
+    BINARY_LOGRAW: [PREDICTED_LABEL, LABELS, RAW_SCORE, RAW_SCORES],
+    BINARY_HINGE: [PREDICTED_LABEL, LABELS, RAW_SCORE, RAW_SCORES],
+    MULTI_SOFTMAX: [PREDICTED_LABEL, LABELS, RAW_SCORE, RAW_SCORES],
+    MULTI_SOFTPROB: [PREDICTED_LABEL, LABELS, PROBABILITY, PROBABILITIES, RAW_SCORE, RAW_SCORES]
+}
+
+
+def is_selectable_inference_response():
+    return SAGEMAKER_INFERENCE_OUTPUT in os.environ
+
+
+def get_selected_content_keys():
+    """Get the selected content keys from the `SAGEMAKER_INFERENCE_OUTPUT` env var.
+
+    :return: selected content keys (list of str)
+    """
+    if is_selectable_inference_response():
+        return os.getenv(SAGEMAKER_INFERENCE_OUTPUT).replace(' ', '').lower().split(',')
+    raise RuntimeError("'SAGEMAKER_INFERENCE_OUTPUT' environment variable is not present. "
+                       "Selectable inference content is not enabled.")
+
+
+def _get_labels(objective, num_class=""):
+    if "binary:" in objective:
+        return [0, 1]
+    if "multi:" in objective and num_class:
+        return list(range(int(num_class)))
+    return np.nan
+
+
+def _get_predicted_label(objective, data):
+    if objective in [BINARY_HINGE, MULTI_SOFTMAX]:
+        return data
+    if objective in [BINARY_LOG]:
```

Comment:
This makes me nervous, since we are now changing the value returned to customers. I don't have intuition on if this is the right behavior to implement. Please explain reasoning.

Reply:
This behavior is the same as the logic in the [sklearn xgboost API](https://xgboost.readthedocs.io/en/release_0.90/python/python_api.html#module-xgboost.sklearn). The `XGBClassifier` has similar logic in its [`.predict(..)`](https://github.com/dmlc/xgboost/blob/release_1.0.0/python-package/xgboost/sklearn.py#L892-L896) which works for all classification objective functions. The code for determining predicted labels is based on that sklearn xgboost wrapper.